File: prep/implementation/patterns_reference_implementation/sliding_window/MaximumSumSubarrayOfSizeK.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

// O(n) / O(1)
public class MaximumSumSubarrayOfSizeK {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();
        int k = in.nextInt();

        int[] a = new int[n];
        for (int i = 0; i < n; i++) {
          a[i] = in.nextInt();
        }

        int maxSum = 0, currSum = 0;
        int windowStart = 0;

        for (int windowEnd = 0; windowEnd < n; windowEnd++) {
          currSum += a[windowEnd];

          if (windowEnd >= k - 1) {
            maxSum = Math.max(maxSum, currSum);
            currSum -= a[windowStart];
            windowStart++;
          }
        }

        System.out.println(maxSum);
      }
    }
  }
}

File: prep/implementation/patterns_reference_implementation/sliding_window/maximum_sum_subarray_of_size_k.cpp
#include <algorithm>
#include <iostream>
#include <string>
#include <vector>

using namespace std;

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, n, k;
  cin >> tt;

  while (tt--) {
    cin >> n >> k;

    vector<int> a(n);
    for (int i = 0; i < n; i++) {
      cin >> a[i];
    }

    int max_sum = 0, curr_sum = 0;
    int window_start = 0;

    for (int window_end = 0; window_end < n; window_end++) {
      curr_sum += a[window_end];

      if (window_end >= k - 1) {
        max_sum = max(max_sum, curr_sum);
        curr_sum -= a[window_start];
        window_start++;
      }
    }

    cout << max_sum << "\n";
  }

  return 0;
}